File: agent/launch_evidence.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from hashlib import sha256

if __package__:
    from .ai_disclosure import load_ai_disclosure
    from .route_manifest import EXPECTED_REVISION as ROUTE_MANIFEST_REVISION
    from .route_manifest import load_route_manifest
else:
    from ai_disclosure import load_ai_disclosure
    from route_manifest import EXPECTED_REVISION as ROUTE_MANIFEST_REVISION
    from route_manifest import load_route_manifest


INDEX_POLICY_REVISION = "index-policy-v1"
RESPONSE_MATRIX_REVISION = "launch-safety-matrix-v1"
CACHE_ISOLATION_REVISION = "launch-cache-isolation-v1"
SITEMAP_PROTOCOL_REVISION = "pinned-sitemap-bundle-v1"

_SHA256 = re.compile(r"^[0-9a-f]{64}$")


def _validated_revision(value: object, label: str) -> str:
    if type(value) is not str:
        raise TypeError(f"{label} must be a string")
    if not value or value.strip() != value:
        raise ValueError(f"{label} must be a non-empty canonical revision")
    return value


def _validated_sha256(value: object, label: str) -> str:
    if type(value) is not str:
        raise TypeError(f"{label} must be a string")
    if _SHA256.fullmatch(value) is None:
        raise ValueError(f"{label} must be a lowercase SHA-256 digest")
    return value


def build_policy_fingerprint(
    *,
    route_revision: str,
    route_digest: str,
    disclosure_revision: str,
    disclosure_digest: str,
) -> str:
    payload = {
        "index_policy": _validated_revision(
            INDEX_POLICY_REVISION, "index policy revision"
        ),
        "response_matrix": _validated_revision(
            RESPONSE_MATRIX_REVISION, "response matrix revision"
        ),
        "cache_isolation": _validated_revision(
            CACHE_ISOLATION_REVISION, "cache isolation revision"
        ),
        "sitemap_protocol": _validated_revision(
            SITEMAP_PROTOCOL_REVISION, "sitemap protocol revision"
        ),
        "route_artifact": {
            "revision": _validated_revision(route_revision, "route revision"),
            "sha256": _validated_sha256(route_digest, "route digest"),
        },
        "disclosure_artifact": {
            "revision": _validated_revision(disclosure_revision, "disclosure revision"),
            "sha256": _validated_sha256(disclosure_digest, "disclosure digest"),
        },
    }
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode(
        "utf-8"
    )
    return sha256(canonical).hexdigest()


@dataclass(frozen=True)
class PolicyEvidence:
    policy_fingerprint: str
    route_manifest_revision: str
    backend_policy_revision: str

    def __post_init__(self) -> None:
        _validated_sha256(self.policy_fingerprint, "policy fingerprint")
        route_revision = _validated_revision(
            self.route_manifest_revision, "route manifest revision"
        )
        if route_revision != ROUTE_MANIFEST_REVISION:
            raise ValueError("route manifest revision is not current")
        backend_revision = _validated_revision(
            self.backend_policy_revision, "backend policy revision"
        )
        if backend_revision != INDEX_POLICY_REVISION:
            raise ValueError("backend policy revision is not current")


def current_policy_evidence() -> PolicyEvidence:
    route = load_route_manifest()
    disclosure = load_ai_disclosure()
    return PolicyEvidence(
        policy_fingerprint=build_policy_fingerprint(
            route_revision=route.revision,
            route_digest=route.artifact.sha256,
            disclosure_revision=disclosure.revision,
            disclosure_digest=disclosure.artifact.sha256,
        ),
        route_manifest_revision=route.revision,
        backend_policy_revision=INDEX_POLICY_REVISION,
    )

File: web-nuxt/server/utils/launch/launchEvidence.ts
import { createHash } from 'node:crypto'

export const INDEX_POLICY_REVISION = 'index-policy-v1'
export const RESPONSE_MATRIX_REVISION = 'launch-safety-matrix-v1'
export const CACHE_ISOLATION_REVISION = 'launch-cache-isolation-v1'
export const SITEMAP_PROTOCOL_REVISION = 'pinned-sitemap-bundle-v1'

const SHA256_PATTERN = /^[0-9a-f]{64}$/u

export interface PolicySemanticRevisions {
  readonly indexPolicy: string
  readonly responseMatrix: string
  readonly cacheIsolation: string
  readonly sitemapProtocol: string
}

function validateRevision(value: unknown, label: string): string {
  if (typeof value !== 'string') throw new TypeError(`${label} must be a string`)
  const normalized = value.normalize('NFC')
  if (!normalized || normalized.trim() !== normalized) {
    throw new Error(`${label} must be a non-empty canonical revision`)
  }
  return normalized
}

function validateSha256(value: unknown, label: string): string {
  if (typeof value !== 'string') throw new TypeError(`${label} must be a string`)
  if (!SHA256_PATTERN.test(value)) throw new Error(`${label} must be a lowercase SHA-256 digest`)
  return value
}

export function buildPolicyFingerprint(
  input: {
    routeRevision: string
    routeDigest: string
    disclosureRevision: string
    disclosureDigest: string
  },
  overrides: Partial<PolicySemanticRevisions> = {},
): string {
  const semantic: PolicySemanticRevisions = {
    indexPolicy: overrides.indexPolicy ?? INDEX_POLICY_REVISION,
    responseMatrix: overrides.responseMatrix ?? RESPONSE_MATRIX_REVISION,
    cacheIsolation: overrides.cacheIsolation ?? CACHE_ISOLATION_REVISION,
    sitemapProtocol: overrides.sitemapProtocol ?? SITEMAP_PROTOCOL_REVISION,
  }

  // Keep key order and compact UTF-8 JSON identical to Python's sorted JSON.
  const canonicalPayload = JSON.stringify({
    cache_isolation: validateRevision(semantic.cacheIsolation, 'cache isolation revision'),
    disclosure_artifact: {
      revision: validateRevision(input.disclosureRevision, 'disclosure revision'),
      sha256: validateSha256(input.disclosureDigest, 'disclosure digest'),
    },
    index_policy: validateRevision(semantic.indexPolicy, 'index policy revision'),
    response_matrix: validateRevision(semantic.responseMatrix, 'response matrix revision'),
    route_artifact: {
      revision: validateRevision(input.routeRevision, 'route revision'),
      sha256: validateSha256(input.routeDigest, 'route digest'),
    },
    sitemap_protocol: validateRevision(semantic.sitemapProtocol, 'sitemap protocol revision'),
  })
  return createHash('sha256').update(canonicalPayload, 'utf8').digest('hex')
}

export function isSha256(value: unknown): value is string {
  return typeof value === 'string' && SHA256_PATTERN.test(value)
}
